tests/unit/resource/suspend_pool.cpp

```diff
@@ -0,0 +1,160 @@
+//  Copyright (c) 2017 Mikael Simberg
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// Simple test verifying basic resource_partitioner functionality.
+
+#include <hpx/hpx_start.hpp>
+#include <hpx/include/async.hpp>
+#include <hpx/include/resource_partitioner.hpp>
+#include <hpx/include/threadmanager.hpp>
+#include <hpx/include/threads.hpp>
+#include <hpx/runtime/threads/executors/pool_executor.hpp>
+#include <hpx/runtime/threads/policies/scheduler_mode.hpp>
+#include <hpx/runtime/threads/policies/schedulers.hpp>
+#include <hpx/runtime/threads/thread_helpers.hpp>
+#include <hpx/util/lightweight_test.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <string>
+#include <utility>
+#include <vector>
+
+int hpx_main(int argc, char* argv[])
+{
+    try
+    {
+        hpx::this_thread::get_pool()->suspend();
+        HPX_TEST_MSG(false, "Suspending is not possible on own pool");
+    }
+    catch (std::runtime_error const&)
+    {
+    }
+
+    hpx::threads::detail::thread_pool_base& worker_pool =
+        hpx::resource::get_thread_pool("worker");
+    hpx::threads::executors::pool_executor worker_exec("worker");
+    std::size_t const worker_pool_threads =
+        hpx::resource::get_num_threads("worker");
+
+    {
+        // Suspend and resume pool
+        hpx::util::high_resolution_timer t;
+
+        while (t.elapsed() < 5)
+        {
+            std::vector<hpx::future<void>> fs;
+
+            for (std::size_t i = 0;
+                 i < worker_pool_threads * 10000; ++i)
+            {
+                fs.push_back(hpx::async(worker_exec, [](){}));
+            }
+
+            worker_pool.suspend();
+
+            // Can only suspend once all work is done
+            auto f = hpx::when_all(std::move(fs));
+            HPX_TEST(f.is_ready());
+
```

Comment:
Is there a way for the `suspend()` API to return a `future<void>` allowing to synchronize with the fact that the pool actually has been suspended?

Reply:
I suppose it could. However, I've again assumed that this can be called from outside the runtime in which case waiting for the future will not work. I don't have a strong case for calling it from outside the runtime but thought it's unnecessary to restrict it unless needed. This may be a reason to restrict it though.

Reply:
Alternatively you could add a callback that is to be called once the operation is finished executing. That also would allow for splitting the API into two functions, one taking the callback (to be used when called from outside HPX), and one (internally) passing a `hpx::lcos::local::packaged_task` to the first API and returning a future extracted from the packaged_task (to be called from HPX threads). That should decouple the waiting from the actual context the suspend is invoked from.

Reply:
Would you then spawn an OS thread or HPX thread depending on from where it's called? The downside is that there will be two functions, and the user has to know (finnd out) that only one of them can be called from outside the runtime (asserts help though, and returning a future is already a good sign it can't be called from the outside). I'll give it a shot in any case as it's a nice idea.

Reply:
I think you can always spawn an HPX thread. Invoking the callback from an HPX thread should be a viable option, always.

Reply:
Ah, you're right. The problem is that the suspending thread will then run on the pool it's supposed to suspend. So either it gets stuck or it has to call the callback before it has fully suspended the pool. The latter feels wrong but could be good enough (it does first wait for all work to be done before really suspending). Maybe there's another way.

Reply:
Two corrections to my previous comments: I have to launch an OS thread in the case of suspending from non-HPX threads, and suspending from non-HPX threads is required for suspending the whole runtime...

I implemented a futurized and a callback version of `suspend` the way you suggested and the principle works well but I still need to tidy it up. The real question is: do you think it's useful that all of `suspend`/`resume`/`suspend_processing_unit`/`resume_processing_unit` have a futurized and a callback version? I think the answer is clearly yes if it's useful for `suspend`.

Reply:
@hkaiser Besides technical details, do you have an opinion on the last question? (In case this slipped past you...)